File: backend/app/routers/genes.py
from fastapi import APIRouter, Query

from app.schemas.gene import GeneSearchResult
from app.services import gene_service

router = APIRouter()


@router.get("/search", response_model=list[GeneSearchResult])
async def search_genes(
    q: str = Query(..., description="Search query"),
    organism: str | None = Query(None, description="Filter by organism"),
    limit: int = Query(20, ge=1, le=100),
):
    return await gene_service.search_genes(q, organism=organism, limit=limit)


@router.get("/{gene_id}")
async def get_gene(gene_id: str):
    return await gene_service.get_gene(gene_id)


@router.get("/{gene_id}/sequence")
async def get_gene_sequence(gene_id: str, seq_type: str = Query("cds")):
    return await gene_service.get_gene_sequence(gene_id, seq_type=seq_type)

File: frontend/src/lib/api/genes.ts
import type { GeneDetail, GeneSearchResult, GeneSequence } from "@/types";
import apiClient from "./client";

export async function searchGenes(query: string, organism?: string): Promise<GeneSearchResult[]> {
  const { data } = await apiClient.get("/api/genes/search", {
    params: { q: query, organism },
  });
  return data;
}

export async function getGene(geneId: string): Promise<GeneDetail> {
  const { data } = await apiClient.get(`/api/genes/${geneId}`);
  return data;
}

export async function getGeneSequence(geneId: string, seqType = "cds"): Promise<GeneSequence> {
  const { data } = await apiClient.get(`/api/genes/${geneId}/sequence`, {
    params: { seq_type: seqType },
  });
  return data;
}
